### PR TITLE
[CP] Make CP loss curve on par with TP

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129517
* #129516
* #130527
* __->__ #129515
* #129514
* #129512

Summary:
This PR changes two implementations to make CP (CP8) lose curve be on par with TP (TP8).

1. Making key and value contiguous before doing ring attention. It is unclear why this is a requirement as SDPA does not have this requirement.

2. Use the out, grad_out, softmax_lse passed by autograd to do the backward. This implementation is similar to the implementation in transformer engine. The original implementation reruns the SDPA to get the output and logsumexp and uses that reculcated results to infer the corrected softmax_lse. But that implementation does not give a better accuracy or lose curve. Instead, that implementation converges slower.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @yf225 @chauhang

### CODE REVIEW

File: torch/distributed/_tensor/experimental/attention.py [hunk 1]
```diff
@@ -231,6 +232,11 @@ def _templated_ring_attention(
 
     chunks = []
     logsumexps = []
+    # Without making key and value contiguous(), the lose curve is bad.
+    # TODO(fegin): figure out why this is a requirement since SDPA does not have
+    # this requirement.
```
Comment (d4l3k) [inline torch/distributed/_tensor/experimental/attention.py:237]: @drisspg should we be checking this in the sdpa operators?

File: torch/distributed/_tensor/experimental/attention.py [hunk 2]
```diff
@@ -411,39 +417,44 @@ def _scaled_dot_product_ring_flash_attention_backward(
...
+                    dropout_p=dropout_p,
+                    is_causal=is_causal_behavior.value,
+                    scale=scale,
+                )
+                softmax_lse_corrected = torch.exp(logsumexp - softmax_lse)
+                grad_out_ = grad_out * softmax_lse_corrected.conj().unsqueeze(-1).to(
+                    grad_out.dtype
+                )
+            else:
+                grad_out_ = grad_out
```
Comment (d4l3k) [inline torch/distributed/_tensor/experimental/attention.py:448]: It's not obvious to me why it's correct to just use grad_out/softmax_lse -- shouldn't the lse be different for each chunk?

I guess using the merged one is potentially better as it's the "true" lse rather than the partially accumulated lse? but still not obvious mathematically it's correct

Can you add a comment with some of the reasoning/docs for why we do it this way?

**Review by d4l3k — APPROVED** [review]

File: torch/distributed/_tensor/experimental/attention.py (hunk 1 above)
Comment (drisspg) [reply torch/distributed/_tensor/experimental/attention.py:237]: We do have many checks within the operators both at the runtime dispatcher, however I see that this coe is not calling the 'can use functions'
https://github.com/pytorch/pytorch/blob/9554a9af8788c57e1c5222c39076a5afcf0998ae/torch/nested/_internal/sdpa.py#L278
https://github.com/pytorch/pytorch/blob/9554a9af8788c57e1c5222c39076a5afcf0998ae/torch/nested/_internal/sdpa.py#L297

As well within the kernel we have last line of dense checks:
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L369-L393

I really dont know why the contigous calls are needed unless there is some other invariant not being satisfied

**Review by wanchaol — APPROVED** [review]

File: torch/distributed/_tensor/experimental/attention.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:448]: Nvidia's implementation also does not recalculate the grad_out. Also, the local verification shows that without recalculation, the gradients are more close to the original SDPA.

Comment (fegin) [reply torch/distributed/_tensor/experimental/attention.py:449]: The original ring attention paper, https://arxiv.org/pdf/2310.01889, also does not perform re-computation (figure 4).
